### ISSUE TITLE
IDE should detect previously generated 'GetHashCode' and offer to upgrade it to use System.HashCode if available.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#38959]

**Comment by CyrusNajmabadi** [comment]
Tagging @jasonmalinowski @sharwell i'm willing to provide a fix here.

**Comment by jnm2** [comment]
@CyrusNajmabadi Would you also detect GetHashCode generated by ReSharper? I'd find that useful but I can picture people wanting one and not the other.

**Comment by CyrusNajmabadi** [comment]
@jnm2 sure, i can try that.  can you give an example of what it generates (ideally with a few reference, nullable and value types)?

**Comment by jnm2** [comment]
@CyrusNajmabadi I can do more too, just off the top of my head

```cs
class Foo
{
    public int IntProperty { get; }
    public bool BoolProperty { get; }
    public bool? NullableBoolProperty { get; }
    public string StringProperty { get; }
    public AttributeTargets EnumProperty { get; }

    public override int GetHashCode()
    {
        unchecked
        {
            var hashCode = IntProperty;
            hashCode = (hashCode * 397) ^ BoolProperty.GetHashCode();
            hashCode = (hashCode * 397) ^ NullableBoolProperty.GetHashCode();
            hashCode = (hashCode * 397) ^ (StringProperty != null ? StringProperty.GetHashCode() : 0);
            hashCode = (hashCode * 397) ^ (int)EnumProperty;
            return hashCode;
        }
    }
}
```

**Comment by CyrusNajmabadi** [comment]
@jnm2 Nope!  that looks great.  Should be able to manage from that!

**Comment by CyrusNajmabadi** [comment]
> This is with:

Can you show me what it generates if you change the string comparison?

**Comment by jnm2** [comment]
```cs
class Foo
{
    public string StringProperty { get; }

    public override int GetHashCode()
    {
        return (StringProperty != null ? StringComparer.CurrentCultureIgnoreCase.GetHashCode(StringProperty) : 0);
    }
}
```

**Comment by sharwell** [comment]
We will need to decide if this API analyzer and code fix should live here, or be implemented alongside dotnet/corefx#40739.

**Comment by CyrusNajmabadi** [comment]
I feel like this should live here.  The IDE is already the one that has been generating this form of GetHashCode (not anyone else).  And we're the ones that already know and use System.HashCode instead if it is available.  The part that is missing is that if the user uses the IDE to generate GEtHashCode in the past, they have no facility to move forward, except deleting what they have and regenerating.

Since this code-gen is entirely something we caused, we need to take responsibility for it.

**Comment by ryzngard** [comment]
Design Review: We will take this change as is, with IDE code targeted to be fixed. We believe a better approach should be encouraged for corefx analyzers to help users with the annotation around hashcode.

**Comment by CyrusNajmabadi** [comment]
Fixed with https://github.com/dotnet/roslyn/pull/38973
